0.13: some updates in "Using Garden" guide series (#4105)
garden-io/garden: docs/using-garden/actions.md

**TODO: Review this and correct/remove if necessary. This seems to be unnecessary for actions.**
{% hint style="warning" %}
Note that you **must** use the `include` and/or `exclude` directives (described above) when module paths overlap. This
is to help users steer away from subtle bugs that can occur when modules unintentionally consume source files from other
modules. See the next section for details on including and excluding files.
{% endhint %}

## Actions in the Stack Graph

The key concept and the main benefit of having _actions_ instead of modules is that **anything can depend on anything**.

Some [prior constraints](./modules.md#modules-in-the-stack-graph), that have tripped people up in the past, have been
dropped. A dependency is then specified as `<kind>.<name>`, e.g. `build.my-image` or `deploy.my-deploy`.
There will still be some specific semantics for different types of dependencies, but quite a bit simpler than before
(which was often difficult to reason about as well).

## Examples

In this section we consider a few very basic examples with configuration code-snippets.

> See more examples in the [examples on our GitHut repository](../../examples).

Reviewer: ```suggestion
See more [examples on our GitHut repository](../../examples).
```

* `Deploy` action to _deploy_ something, e.g. a built and configured application container. It is a replacement for old module-based [services](./services.md) configuration.
* `Run` action to _run_ something, e.g. some custom scripts. It is a replacement for old module-based [tasks](./tasks.md) configuration.
* `Test` action to _test_ something, e.g. run a single application's test suite. It is a replacement for old module-based [test](./tests.md) configuration.

Generally, a Garden project may contain multiple _components_ like applications, databases, etc.
Each component can have multiple actions. For example, a single application can have different build and deployment
scenarios (`Build` snd `Deploy` actions), multiple test suites (`Test` actions), and various helper scripts (`Run` actions).

Each action can be defined in own Garden configuration file. But, for the sake of simplicity and maintainability, it
might be easier to have one `garden.yml` configuration file per project's component.

Below is a simple example of the action configurations of the `backend` application (from

> the [`demo-project`](../../examples/demo-project) example project):

Reviewer: I'm not 100% sure if this will work with gitbook. All other examples are marked with absolute github links. If it does that would be great though.

Author: This should work fine. I used relative links in the local mode guide, and it worked well in docs.garden.io.

```yaml
kind: Build
name: backend
description: Backend app container image
type: container

---

kind: Deploy
name: backend
description: Backend app container
type: container

build: backend

# You can specify variables here at the action level
variables:
  ingressPath: /hello-backend

spec:
  healthCheck:
    httpGet:
      path: /hello-backend
      port: http
  ports:
    - name: http
      containerPort: 8080
      servicePort: 80
  ingresses:
    - path: ${var.ingressPath}
      port: http

---

kind: Run
name: backend-test
type: container
build: backend
spec:
  command: [ "sh", "-c", "echo task output" ]
```

The action configuration style above resembles the old-fashioned [module-based](./modules.md) configuration
where [services](./services.md), [tasks](./tasks.md), and [tests](./tests.md) were configured in the same file. Such
configuration style pattern makes the migration to `0.13` easier and faster.

## How it Works

> A Garden project is usually split up into the project-level configuration file, and several action-level configuration

Reviewer: ```suggestion
A Garden project is usually split up into the project-level configuration file and several action-level configuration
```

files, each in the root directory of the respective part of the project:

```console
.
├── project.garden.yml
├── application-a
│   ├── actions.garden.yml
│   └── ...
├── application-b
│   └── actions.garden.yml
│   └── ...
├── application-c
     └── actions.garden.yml
     └── ...
```

> You can also choose any `*.garden.yml` filename for each configuration file.

Reviewer: "choose any" sounds like there's a limited selection to choose from. I'd recommend to rewording to something like this
```suggestion
Filenames ending with`.garden.yml` will be picked up by Garden.
```

For example, you might prefer to set the application name in the filename, e.g. `application-a.actions.garden.yml` to
make it easier to find in a large project.

Alternatively, you can have the actions grouped by [`kind`](#action-kinds) in a dedicated config file,
i.e. `deploy-actions.garden.yml`, `test-actions.garden.yml`, and so on.

Another option is to have each action defined in a separate configuration file. It might be not convenient if you have a
large number of actions in your project.

Read the sections below for more information on configuring actions, including how to control which files and
directories are included in a single action.

### Action types

Each [`kind`](#action-kinds) of action must have a _type_. Different action _types_ behave in different ways.
For example, the `container` type of the `Build` action corresponds to a Docker image, either built from a

> local `Dockerfile` or pulled from a remote repository.

Reviewer: This line is false. A build action is not needed if an already built image from a repository is used.

Author: So, does it behave anyhow differently from module-based builds?

What the line above says is that the image can be either built locally or pulled from a remote repository. The `Build` action declaration is still necessary, isn't it? 

Let's consider an example:
```yaml
kind: Build
name: postgres-db
description: Postgres DB container
type: container
image: postgres:12 # pulls the image from a remote Docker repo, no local build happens

---

kind: Deploy
name: postgres-db
build: postgres-db # refers to the name of the Build action above
...
```

We can't use `build: postgres:12` in the `Deploy` action above because there is no `Build` action named `postgres:12` defined in the config, can we? Please, correct me if I'm wrong.

Thus, there is nothing to build locally, but the `Build` action is still executed - it either finds an already existing pulled image in the local Docker registry (and this is a pure no-op "build") or pulls a new one from the remote repository.

Reviewer: There's no `image` field on a container build action. If you want to deploy an existing image you set it under the `spec.image` field of a container deploy. In that case a build does not have to be specified. https://github.com/garden-io/garden/blob/0.13/docs/reference/action-types/Deploy/container.md#specimage

Author: Ah, right! Thank you! 👍 I'll fix this.

Author: Deleted that obsolete snippet in https://github.com/garden-io/garden/pull/4105/commits/1097e3a9917466e6ee754a26f464d34a0b384635

We have a reference to the action types page.

  flexible templating capabilities, or if you aim to publish the charts.
- [exec](../reference/action-types/Run/exec.md) `Run` actions offer a flexible way to weave in arbitrary scripts and
  commands that are executed locally. These can be custom build steps, scripts, tests or really anything else. The
  caveat is that they always run on the same machine as the Garden CLI, and not e.g. in a Kubernetes cluster, and thus
  not quite as portable.
- [terraform](../reference/action-types/Deploy/terraform.md)  offer a powerful way to deploy any cloud resources as part
  of your project. See the [Terraform guide](../terraform-plugin/README.md) for more information.

There are several other action types available as well. See
the [action types reference](../reference/action-types/README.md) for a full list of supported action types, and their
configuration reference.

### Including and excluding files

By default, all files in the same directory as an action configuration file are included as source files for that
action.
Sometimes you need more granular control over the context, not least if you have multiple actions in the same directory.

> The `include` and `exclude` fields are a simple way to explicitly specify which sources should belong to a particular

Reviewer: ```suggestion
The `include` and `exclude` fields are used to explicitly specify which sources should belong to a particular
```

We try to avoid words as "easy" or "simple" as for the user it might not seem that way

> action. They both accept a list of POSIX-style paths or globs. For example:

Reviewer: ```suggestion
action. Both accept a list of POSIX-style paths or globs. For example:
```
or
```suggestion
action. Both of them accept a list of POSIX-style paths or globs. For example:
```

```yaml
kind: Build
description: My container
type: container
include:
  - Dockerfile
  - my-sources/**/*.py
exclude:
  - my-sources/tmp/**/*
...
```

> Here, we only include the `Dockerfile` and all the `.py` files under `my-sources/`, but exclude the `my-sources/tmp`

Reviewer: ```suggestion
Here we only include the `Dockerfile` and all the `.py` files under `my-sources/`, but exclude the `my-sources/tmp`
```

directory.

If you specify a list with `include`, only those files/patterns are included. If you then specify one or more `exclude`
files or patterns, those are filtered out of the files matched by `include`. If you _only_ specify `exclude`, those
patterns will be filtered out of all files in the action directory.

Note that the action `include` and `exclude` fields have no effect on which paths Garden watches for changes. Use
the [project `scan.exclude` field](./projects.md) for that purpose.

You can also use [.gardenignore file](./configuration-overview.md#ignore-file), much like `.gitignore` files, to exclude
files across your project. You can place them in your project root, in action roots, and even in individual
sub-directories of actions.

> ### Multiple actions in the same directory

Reviewer: I'm not sure if this section is needed. All the necessary info was already [written earlier](https://github.com/garden-io/garden/pull/4105/files#diff-778425512afdaa87f3cc1f68b8dd4053c2efc6a6ac0e5da6c1940cbc5251babdR99-R103)

Author: Agree, I was thinking about combining those sections. I got both of them from the original modules guide. I'll change this soon.

Author: Remove the section in https://github.com/garden-io/garden/pull/4105/commits/f3f67ef67abd6d15974599d158ab1be1b9438081 as it didn't add anything new.

> Sometimes, it's useful to define several actions in the same `garden.yml` file. One common situation is where more than

Reviewer: ```suggestion
Sometimes it's useful to define several actions in the same `garden.yml` file. One common situation is where more than
```

one `Dockerfile` is in use (e.g. one for a development build and one for a production build). You may only have a
handful of actions, and it may be the cleanest approach to define all of them in `actions.garden.yml` in your project
root.

> It also might be better for maintainability to have multiple actions of the same `kind` in the same configuration file,

Reviewer: This point is repeated on L99

Author: Removed it

dependencies:
  - test.b-integration-testing-backend
spec:
  args: [ npm, run, integ-b ]
```

[//]: # (TODO: Review this and correct/remove if necessary. This seems to be unnecessary for actions.)
**TODO: Review this and correct/remove if necessary. This seems to be unnecessary for actions.**
{% hint style="warning" %}
Note that you **must** use the `include` and/or `exclude` directives (described above) when module paths overlap. This
is to help users steer away from subtle bugs that can occur when modules unintentionally consume source files from other
modules. See the next section for details on including and excluding files.
{% endhint %}

## Actions in the Stack Graph

The key concept and the main benefit of having _actions_ instead of modules is that **anything can depend on anything**.

Some [prior constraints](./modules.md#modules-in-the-stack-graph), that have tripped people up in the past, have been
dropped. A dependency is then specified as `<kind>.<name>`, e.g. `build.my-image` or `deploy.my-deploy`.
There will still be some specific semantics for different types of dependencies, but quite a bit simpler than before
(which was often difficult to reason about as well).

> ## Examples

Reviewer: maybe add a section about example projects here as well.
